File: 0045-jump-game-ii/0045-jump-game-ii.py
class Solution:
    def jump(self, nums: List[int]) -> int:
        n: int = len(nums)
        if (n <= 1):
            return 0

        max_reachable: int = 0
        jump_count_track: Dict[int,int] = {
            max_reachable:0 # max reachable:jump count needed
        }

        for i, jump in enumerate(nums):
            cur_max_reachable = i + jump
            if cur_max_reachable > max_reachable :
                for max_reachable_i in sorted(jump_count_track.keys()):
                    if (max_reachable_i >= i):
                        jump_count_track[cur_max_reachable] = jump_count_track[max_reachable_i] + 1
                        break

                max_reachable = cur_max_reachable

            if max_reachable >= n - 1:
                return jump_count_track[max_reachable]
        
        return -1
        

File: 0045-jump-game-ii/0045-jump-game-ii.java
class Solution {
    public int jump(int[] nums) {
        int n = nums.length;
        if (n <= 1) {return 0;}
        
        int prevReachable = -1;
        int prevJumpCount = 0;
        
        int maxReachable = nums[0];
        int maxJumpCount = prevJumpCount + 1;

        int i = 1;
        while (i < n && maxReachable < n-1){
            if (i > prevReachable){
                prevReachable = maxReachable;
                prevJumpCount = maxJumpCount;
            }

            if (i + nums[i] > maxReachable){
                maxReachable = i + nums[i];
                maxJumpCount = prevJumpCount + 1;
            }
            System.out.print(i);
            System.out.println(" =======");
            System.out.println(prevReachable);
            System.out.println(prevJumpCount);
            System.out.println(maxReachable);
            System.out.println(maxJumpCount);
            System.out.println();
            i++;

        }
        return maxJumpCount;
    }
}